text/0000-introduce empty ranges.md

```diff
@@ -0,0 +1,425 @@
+
+```
+not_empty = progress and min.finite() and step.finite() // .finite() => the value is not `+-Inf` and not `NaN`
+infinite = not_empty and max.infinite()
+empty = not not_empty
+```
+
+This can be implemented, replacing the above original code portion, as follows:
+
+```pony
+class Range[A: (Real[A] val & Number) = USize] is Iterator[A]
+  ..
+  let _min: A
+  let _max: A
+  let _inc: A
+  let _forward: Bool
+  let _infinite: Bool
+  let _empty: Bool
+  var _idx: A
+
+  new create(min: A, max: A, inc: A = 1) =>
+    _min = min
+    _max = max
+    _inc = inc
+    _forward = (_min < _max) and (_inc > 0)
+    (let min_finite, let max_finite, let inc_finite) =
+      iftype A <: FloatingPoint[A] then
+        (_min.finite(), _max.finite(), _inc.finite())
+      else
+        (true, true, true)
+      end
+    let progress = ((_min < _max) and (_inc > 0))
+                    or ((_min > _max) and (_inc < 0)) // false if any is NaN!
+    if progress and min_finite and inc_finite then
+      _empty = false
+      _infinite = not max_finite // ok to use not max_finite for max_infinite
+                                 // since progress excludes _max == nan
+      _idx = _min
+    else
+      _empty = true
+      _infinite = false
+      _idx = _max // has_next() will return false without code modification
+    end
+```
+
+Importantly, the original code for `.has_next()` and `.next()` does not require any changes! A modification that conserves the implementation "trick" to initialize `_idx = _max` for empty Ranges is required to `.rewind()`. Also, analogous to the existing `.is_infinite(): Bool` function, there is also a new `.is_empty()`.
+
+```
+  fun ref rewind() =>
+    if _empty then _idx = _max else _idx = _min end // currently _idx = _min
+
+  fun is_empty(): Bool =>
+    _empty
+```
+
+## Examples of Affected Ranges
+
+Here is a list of Range examples all of which were previously infinite, and their new classification:
+
+**Empty Ranges**
+
+```
+Range(0, 10, -1), .is_empty() == true, .is_infinite() == false
+Range(10, 0, 1), .is_empty() == true, .is_infinite() == false
+Range(10, 10, 1), .is_empty() == true, .is_infinite() == false
+Range(10, 10, -1), .is_empty() == true, .is_infinite() == false
+Range(10, 10, 0), .is_empty() == true, .is_infinite() == false
+Range(-10, -10, 1), .is_empty() == true, .is_infinite() == false
+Range(-10, -10, -1), .is_empty() == true, .is_infinite() == false
+Range(-10, -10, 0), .is_empty() == true, .is_infinite() == false
+Range(0, 10, -1).is_empty() == true, .is_infinite() == false
+Range(10, 0, 1).is_empty() == true, .is_infinite() == false
+Range(10, 10, 1).is_empty() == true, .is_infinite() == false
+Range(10, 10, -1).is_empty() == true, .is_infinite() == false
+Range(10, 10, 0).is_empty() == true, .is_infinite() == false
```

Comment:
Some of these ranges are repeated such as `Range(0, 10 -1)` on lines 202 and again on 210.

Reply:
Yes. c&p errors that need to be fun xed

Reply:
Been there!